ARROW-10031: [CI][Java] Support Java benchmark in Archery

kiszk opened the pull request:
This PR supports Java benchmark in Ursabot. The implementation is based on [this suggestion](https://mail-archives.apache.org/mod_mbox/arrow-dev/202008.mbox/%3cCABNn7+q35j7QWsHJBX8omdewKT+F1p_M7r1_F6szs4dqc+Luyg@mail.gmail.com%3e)

Here are work items.
- [x] Support `--language=[cpp|java]` option in `diff`
- [x] Enable to build java binding
- [x] Enable to run Java benchmarks
- [x] Allows us to filter/select benchmarks
- [x] Enable to collect results
- [x] Apply the same changes to `run` and `list`

liyafan82 commented:
@kiszk Thank you for doing this. 
Please note that when running the benchmarks, some flags should be configured properly.
They can be set through environmental variables:

ARROW_ENABLE_UNSAFE_MEMORY_ACCESS = true
ARROW_ENABLE_NULL_CHECK_FOR_GET = false

or through system properties:

arrow.enable_unsafe_memory_access = true
arrow.enable_null_check_for_get = false

kiszk commented:
@liyafan82 Thank you for your comment. I will set these two properties as default.

kiszk commented:
At my end, I can generate the following JSON file by `archery benchmark diff --language=java ...`

@liyafan82 any comments regarding format and parameters are appreciated.

```
                                                  benchmark      baseline     contender  change %                                                                                                                                                                                                           counters
0      org.apache.arrow.vector.IntBenchmarks.setIntDirectly  22.500 us/op  11.209 us/op   -50.182  {'mode': 'avgt', 'threads': 1, 'warmups': 5, 'warmupTime': '10 s', 'measurements': 4, 'measurementTime': '10 s', 'jvmArgs': ['-Darrow.enable_null_check_for_get=false -Darrow.enable_unsafe_memory_access=true']}
2  org.apache.arrow.vector.IntBenchmarks.setWithValueHolder  19.031 us/op   6.627 us/op   -65.179  {'mode': 'avgt', 'threads': 1, 'warmups': 5, 'warmupTime': '10 s', 'measurements': 4, 'measurementTime': '10 s', 'jvmArgs': ['-Darrow.enable_null_check_for_get=false -Darrow.enable_unsafe_memory_access=true']}
1       org.apache.arrow.vector.IntBenchmarks.setWithWriter  32.626 us/op  10.246 us/op   -68.594  {'mode': 'avgt', 'threads': 1, 'warmups': 5, 'warmupTime': '10 s', 'measurements': 4, 'measurementTime': '10 s', 'jvmArgs': ['-Darrow.enable_null_check_for_get=false -Darrow.enable_unsafe_memory_access=true']}
```

liyafan82 commented:
> At my end, I can generate the following JSON file by `archery benchmark diff --language=java ...`
> 
> @liyafan82 any comments regarding format and parameters are appreciated.
> 
> ```
>                                                   benchmark      baseline     contender  change %                                                                                                                                                                                                           counters
> 0      org.apache.arrow.vector.IntBenchmarks.setIntDirectly  22.500 us/op  11.209 us/op   -50.182  {'mode': 'avgt', 'threads': 1, 'warmups': 5, 'warmupTime': '10 s', 'measurements': 4, 'measurementTime': '10 s', 'jvmArgs': ['-Darrow.enable_null_check_for_get=false -Darrow.enable_unsafe_memory_access=true']}
> 2  org.apache.arrow.vector.IntBenchmarks.setWithValueHolder  19.031 us/op   6.627 us/op   -65.179  {'mode': 'avgt', 'threads': 1, 'warmups': 5, 'warmupTime': '10 s', 'measurements': 4, 'measurementTime': '10 s', 'jvmArgs': ['-Darrow.enable_null_check_for_get=false -Darrow.enable_unsafe_memory_access=true']}
> 1       org.apache.arrow.vector.IntBenchmarks.setWithWriter  32.626 us/op  10.246 us/op   -68.594  {'mode': 'avgt', 'threads': 1, 'warmups': 5, 'warmupTime': '10 s', 'measurements': 4, 'measurementTime': '10 s', 'jvmArgs': ['-Darrow.enable_null_check_for_get=false -Darrow.enable_unsafe_memory_access=true']}
> ```

@kiszk Thanks for your effort. Generally, it looks great! Some minor comments:
1. It is clearer to rename title 'counters' to 'configuration'?
2. I am curious how are the benchmarks sorted, by 'change %'?

kiszk commented:
For 1., I will rename the title for cpp and Java.

For 2, you are right as sorted [here](https://github.com/apache/arrow/blob/master/dev/archery/archery/cli.py#L595).

kiszk commented:
@liyafan82 @fsaintjacques @kszucs Would it be possible to review this?

kiszk commented:
The following commands should work:
```
archery benchmark list --langauge=java
archery benchmark run --langauge=java
archery benchmark diff --langauge=java
```

Here is an example
```
archery benchmark run  ARROW-10031  --output=out-java.json --language=java  --benchmark-filter=IntBench --java-options=-mx1g --build-extras=-Dfoo.bar=1 --benchmark-extras=-Dbar.foo=2
````

kiszk commented on java/performance/pom.xml:
```diff
@@ -169,10 +173,17 @@
+                        <argument>${benchmark.list}</argument>
+                        <argument>-rff</argument>
+                        <argument>${benchmark.resultfile}</argument>
+                        <argument>-rf</argument>
```
@liyafan82 Here is a change from #8245 . Since I cannot find to add `-rf json` or not to add it in `<arguments>`, the current implementation always generates `jmh-result.json` and change a file name if it will be overridden.

I would like to update here if there is a selection method to add `-rf json` or not to add it.

liyafan82 replied:
It looks good. Thanks.

kiszk commented:
ping @liyafan82 @fsaintjacques @kszucs

liyafan82 commented:
The Java changes look good to me. However, I am not farmiliar with the archery code. So @fsaintjacques @kszucs could you please take a look?

kszucs commented:
Thanks for working on this! At the moment we can only try this out locally since we no longer maintain the buildbot/ursabot setup. I have limited time to review this, so I'd say that if the benchmarking using archery works locally then is should be good to go.

kiszk commented:
I see.

kiszk commented:
@bkietz @xhochy May I ask to review this if possible?

xhochy commented:
Leaving this for @bkietz , `archery` is not my area of expertise.

bkietz commented on dev/archery/archery/lang/java.py:
```diff
@@ -28,3 +31,48 @@ def __init__(self, jar, *args, **kwargs):
+        self.java_home = java_home
+        self.java_options = java_options
+
+        self.build_extras = build_extras
+        self.benchmark_extras = benchmark_extras
+
+    @property
+    def build_definitions(self):
+        extras = list(self.build_extras) if self.build_extras else []
+        return extras
```
I'd recommend doing the cast-to-list here; if an iterator were passed for `build_extras` then it'd be depleted after the first access to `build_definitions`:
```suggestion
        self.build_extras = list(build_extras) if build_extras else []
        self.benchmark_extras = list(benchmark_extras) if benchmark_extras else []

    @property
    def build_definitions(self):
        return self.build_extras
```

I see that CppConfiguration has this same flaw; it's not necessary for you to address it there too

bkietz commented on dev/archery/archery/cli.py:
```diff
@@ -450,19 +511,57 @@ def benchmark_run(ctx, rev_or_path, src, preserve, output, cmake_extras,
+
+@benchmark.command(name="run_cpp")
+@click.pass_context
+def benchmark_run_cpp(ctx, rev_or_path, src, preserve, output, cmake_extras,
```
This division also seems unnecessary; CppBenchmarkRunner and JavaBenchmarkRunner seem similar enough to share more code here

bkietz commented on dev/archery/archery/cli.py:
```diff
@@ -550,19 +649,34 @@ def benchmark_diff(ctx, src, preserve, output, cmake_extras,
+
+@benchmark.command(name="diff_cpp")
+@click.pass_context
+def benchmark_diff_cpp(ctx, src, preserve, output, language, cmake_extras,
```
Again, why separate the languages' benchmarks? It seems especially valuable to have a clearly unified strategy for comparison of benchmarks

kiszk replied:
This separation comes from `@click.pass_context`. To use different options between two languages (e.g. `java-home` or `cxx-flags`), my understanding was that we need different methods with the `@click.pass_context`.

Let me try to have one method with @click.pass_context by lazily parse arguments for each language.

bkietz commented on dev/archery/archery/benchmark/runner.py:
```diff
@@ -205,3 +176,133 @@ def suites(self):
+        run in a file or a string.
+
+        Second, it checks if `rev_or_path` points to a valid Maven build
+        directory.  If so, it creates a JavaenchmarkRunner with this existing
```
```suggestion
        directory.  If so, it creates a JavaBenchmarkRunner with this existing
```
Again, seems like this code didn't need to be repeated

bkietz reviewed (changes requested):
In general, I'd recommend refactoring for less repetition of code between the classes for handling java and C++ builds

kiszk commented on dev/archery/archery/benchmark/jmh.py:
```diff
@@ -0,0 +1,165 @@
+        super().__init__(name, unit, less_is_better, values)
+
+    def __repr__(self):
+        return "JavaMicrobenchmarkHarness[name={},runs={}]".format(
```
[JMH](https://github.com/openjdk/jmh) stands for "Java Microbenchmark Harness". Thus, I used JavaMicrobenchmarkHarness. As you suggested, I will simplify the name as "JavaMicroBench".

emkornfield commented:
@kiszk @bkietz are there more revisions needed here? (it looks like this at least needs a rebase?)

kiszk commented:
@emkornfield @bkietz Thank you for kind ping. Yes, I need an update since I was swamped last month. I will update this PR this week.

emkornfield commented:
CC @dianaclarke who I think has also been working on continuous benchmarking.

kiszk commented:
@emkornfield @bkietz @dianaclarke @liyafan82  I am very sorry for being late to address review comments.

Now, I addressed all of the comments and rebased with the master. Would it be possible to review this PR again?

kszucs commented:
@ursabot --help

kszucs commented:
@ursabot please benchmark lang=C++

kszucs commented:
I wouldn't be too pedantic here, we can address any issues later. Nice addition, thanks @kiszk!

I submitted a conbench build to check that the archery benchmark command remained compatible.

emkornfield commented:
Should this be merged.

kiszk commented:
@bkietz @dianaclarke @liyafan82 Would you have additional comments?

liyafan82 commented:
> @bkietz @dianaclarke @liyafan82 Would you have additional comments?

I have no more comments. Hope this PR will be merged soon.

kou reviewed (approved):
+1

If wee need more refactoring, we can work on it as a follow-up task.

kiszk commented:
Thanks very much for your comments
